service/glacier/internal/customizations/treehash.go

```diff
@@ -0,0 +1,144 @@
+package customizations
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"github.com/awslabs/smithy-go"
+	"github.com/awslabs/smithy-go/middleware"
+	smithyhttp "github.com/awslabs/smithy-go/transport/http"
+	"io"
+)
+
+const bufsize = 1024 * 1024
+
+// AddTreeHashMiddleware adds middleware needed to automatically
+// calculate Glacier's required checksum headers.
+func AddTreeHashMiddleware(stack *middleware.Stack) {
+	stack.Finalize.Add(&TreeHashMiddleware{}, middleware.Before)
+}
+
+// TreeHashMiddleware provides the middleware that will automatically
+// set the sha256 and tree hash headers if they have not already been
+// set.
+type TreeHashMiddleware struct{}
+
+// ID returns the middleware ID.
+func (*TreeHashMiddleware) ID() string { return "Glacier:TreeHash" }
+
+// HandleFinalize implements the finalize middleware handler method
+func (*TreeHashMiddleware) HandleFinalize(
+	ctx context.Context, input middleware.FinalizeInput, next middleware.FinalizeHandler,
+) (
+	output middleware.FinalizeOutput, metadata middleware.Metadata, err error,
+) {
+	req, ok := input.Request.(*smithyhttp.Request)
+	if !ok {
+		return output, metadata, &smithy.SerializationError{
+			Err: fmt.Errorf("unknown request type %T", input.Request),
+		}
+	}
+
+	if err := addChecksum(req); err != nil {
+		return output, metadata, &smithy.SerializationError{Err: err}
+	}
+
+	return next.HandleFinalize(ctx, input)
+}
+
+func addChecksum(req *smithyhttp.Request) error {
+	if req.GetStream() == nil || req.Header.Get("X-Amz-Sha256-Tree-Hash") != "" {
+		return nil
+	}
+
+	h := computeHashes(req.GetStream())
+	if err := req.RewindStream(); err != nil {
+		return err
+	}
```

Comment:
Probably need to check before computing the hash if the stream is seekable or not, and fail when the reader is not seakable with a descriptive error.

To do this smithy-go should expose an accessor to check if the stream is seekable,

https://github.com/awslabs/smithy-go/blob/dab44aa9975953da6d6c9b6ab922800fb6b3be79/transport/http/request.go#L17

Reply:
https://github.com/awslabs/smithy-go/pull/182